Remove some volatile use on objects in corelib

stephentoub opened the pull request:
From a review of usage sites, all of this use for lazy initialization now appears to be superfluous, given our documented memory model.

jkotas reviewed (approved):
Is a new analyzer or something like that motivating this cleanup? Or is this just something that you have run into?

Personally, I like `volatile` as a documentation for lock-free code. I do not have a problem with deleting it where it is not needed.

stephentoub commented:
It's motivated by @vsadov citing earlier this week that such usage still results in extra fences being emitted on arm. If that's not the case, or if we can teach the jit somehow they can be elided, I agree it's better to leave them.

xtqqczze commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs:
```diff
@@ -1141,35 +1141,15 @@ public static CultureInfo GetCultureInfo(string name, bool predefinedOnly)
-            }
-        }
+        private static Dictionary<string, CultureInfo> CachedCulturesByName =>
+            s_cachedCulturesByName ??
```
Does this introduce a race condition where the property could return `null` unexpectedly?

https://github.com/dotnet/runtime/blob/776ab3cd89e1622fe767710f021ab1e86cbdeba3/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs#L782

stephentoub replied:
Yes, if it can be nulled out concurrently, after this compare exchange sets it, it's possible that could cause this to return null.

jkotas replied:
The existing code does not have this race condition. It sounds like this refactoring should be reverted.

stephentoub replied:
Yes, I plan to fix it, I'm just not near my computer that.

stephentoub replied:
Fixed

xtqqczze commented:
@MihuBot

VSadov commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:
```diff
@@ -411,7 +410,7 @@ internal sealed partial class CultureData
         // Cache of regions we've already looked up
         private static volatile Dictionary<string, CultureData>? s_cachedRegions;
-        private static volatile Dictionary<string, string>? s_regionNames;
+        private static Dictionary<string, string>? s_regionNames;
```
Why `s_cachedRegions` is still volatile? Is accessing it needs to be ordered with accessing something else?

VSadov replied:
From the use it looks like just a shared cache. We use a lock when querying/adding to the cache, since it is an ordinary Dictionary. 
We initialize the static without any locks though in a "last to assign wins" fashion. Which makes sense if the cache itself has no identity whatsoever and can be cheaply refilled. It seems doubtful that anyone cares about the order of accesses to `s_cachedRegions`.

stephentoub replied:
I'd only focused on the obvious ones. We can incrementally remove more.

VSadov replied:
There is an interesting scenario when we set `s_cachedRegions = null` in `ClearCachedData`, I assume when somehting changed machine-wide.
If that runs concurrently with a thread doing `s_cachedRegions = tempHashTable`, we could see that installing a cache with data from before the `ClearCachedData`.

I am not familiar with what is happening here at the high level. Perhaps this clearing is coordinated in a way that this kind of race cannot happen...   
Either way `volatile` will not help with that.

VSadov commented on src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:
```diff
@@ -100,7 +101,8 @@ public static class LazyInitializer
         public static T EnsureInitialized<T>([NotNull] ref T? target, Func<T> valueFactory) where T : class =>
-            Volatile.Read(ref target!) ?? EnsureInitializedCore(ref target, valueFactory);
+            target ??
+            EnsureInitializedCore(ref target, valueFactory);
```
I see what you meant in your question about volatile reads inside the Lazy init - what if the initializer does some random writes unrelated to the instance, can we see uninitialized data.

Basically if T's constructor initializes static s, otherwise not related to T.

```cs
if (instT != null) s.ToString();   // no guarantees, can read s before instT

if (EnsureInitialized(ref instT) != null) s.ToString();   // seems like s will be initialized
```

I do not think we document `EnsureInitialized` as an acquiring read, but implementation is an acquiring read. 
Relaxing that might be observable. 

Is this API old enough that someone might depend on this being an acquire?

VSadov replied:
After thinking about it, perhaps the `Volatile.Read` here and `Volatile.Write` in EnsureInitializedCore are intentional - to include all sideeffects of constructor/factory under "initialized" state? 

Especially in the factory case - I can imagine the factory to do a lot of things and then just assign a trivial object as a marker that initialization is done.

VSadov replied:
@stephentoub - just wonder what the final thinking on this. Are we ok with relaxing the guarantees to just the data reachable via the instance?

VSadov replied:
Somehow it was showing to me that the change is still there. Now it looks reverted.

stephentoub committed:
Remove some volatile use on objects in corelib

From a review of the use, all of this use for lazy initialization now appears to be superfluous, given our documented memory model.

VSadov reviewed (approved):
LGTM!

xtqqczze commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:
```diff
@@ -150,7 +149,7 @@ internal sealed partial class CultureData
         private static Dictionary<string, string> RegionNames =>
             s_regionNames ??=
-            new Dictionary<string, string>(257 /* prime */, StringComparer.OrdinalIgnoreCase)
+            new Dictionary<string, string>(255, StringComparer.OrdinalIgnoreCase)
```
https://github.com/dotnet/runtime/pull/33834#discussion_r395926999
